src/TON/Address.cpp

```diff
@@ -0,0 +1,210 @@
+
+using namespace TW;
+using namespace TW::TON;
+using namespace std;
+
+bool Workchain::isValid(WorkchainId_t workchainId) {
+    return (workchainId == MasterChainId || workchainId == BasicChainId);
+}
+
+Address::Address(const std::string& address) {
+    bool valid = parseAddress(address, *this);
+
+    // Ensure address is valid
+    if (!valid) {
+        throw std::invalid_argument("Invalid address data");
+    }
+}
+
+Address::Address(const PublicKey& publicKey) {
+    // Steps: a StateInit account state cell is created (containing code and data), its hash is taken,
+    // and new address is derived from the hash
+
+    if (publicKey.type != TWPublicKeyTypeED25519) {
+        throw std::invalid_argument("Invalid public key type");
+    }
+
+    Cell stateInit = Contract::createStateInit(publicKey);
+
+    // compute hash of stateInit
+    auto hash = stateInit.hash();
+
+    // fill members
+    workchainId = Workchain::defaultChain();
+    addrBytes = hash;
+    isBounceable = true;
+    isTestOnly = false;
+}
+
+bool Address::isValid(const std::string& address) {
+    Address addr;
+    bool isValid = parseAddress(address, addr);
+    if (addr.workchainId != Workchain::defaultChain()) {
+        // not the right chain, refuse
+        return false;
+    }
+    return isValid;
+}
+
+bool Address::parseAddress(const std::string& addressStr_in, Address& addr_inout) {
+    // try several formats, start with the common one, stop if one matches
+    bool isValidUser = parseUserAddress(addressStr_in, addr_inout);
+    if (isValidUser) {
+        return true;
+    }
+    bool isValidRaw = parseRawAddress(addressStr_in, addr_inout);
+    return isValidRaw;
+}
+
+bool Address::parseRawAddress(const std::string& addressStr_in, Address& addr_inout) {
+    // split by colon ':'
+    auto colidx = addressStr_in.find(':');
+
+    if (colidx == std::string::npos) {
+        // no colon, invalid
+        return false;
+    }
+    if (colidx < 1 || colidx >= addressStr_in.length() - 1) {
+        // colon in wrong position
+        return false;
+    }
+    std::string workchainStr = addressStr_in.substr(0, colidx);
+    std::string addressStr = addressStr_in.substr(colidx + 1, addressStr_in.length() - colidx - 1);
+
+    WorkchainId_t workchainId;
+    try {
+        workchainId = std::stoi(workchainStr);
+    } catch (const std::exception& e) {
+        // workchain ID is invalid (not a decimal number)
+        return false;
+    }
+    if (!Workchain::isValid(workchainId)) {
+        // invalid workchain ID
+        return false;
+    }
+    addr_inout.workchainId = workchainId;
+
+    if (addressStr.length() != AddressLength * 2) {
+        // wrong length of address part
+        return false;
+    }
+
+    addr_inout.addrBytes = parse_hex(addressStr);
+
+    addr_inout.isBounceable = true;
+    addr_inout.isTestOnly = false;
+
+    return true;
+}
+
+bool Address::parseUserAddress(const std::string& addressStr_in, Address& addr_inout) {
```

Comment:
maybe `parseBase64Address`?

Reply:
The TON terminology is User and Raw format.